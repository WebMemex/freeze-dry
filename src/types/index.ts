import { UrlString, Fetchy, FrameElement } from './util'
import { SubresourceLink } from '../extract-links/types'
import { Resource } from '../resource'

export * from './util'

// The callback that is run for each encountered subresource link
export type ProcessSubresourceCallback = (
    link: SubresourceLink,
    recurse: ProcessSubresourceRecurse,
) => void | Promise<void>

export type ProcessSubresourceRecurse = (
    link: SubresourceLink,
) => void | Promise<void>

export type NewUrlForResourceCallback = (resource: Resource) => string | Promise<string>

export type ContentSecurityPolicy = string | {
    [directive: string]: string | string[] | undefined | null,
}

export interface GlobalConfig {
    timeout: number,
    signal?: AbortSignal,
    docUrl?: UrlString,
    charsetDeclaration: string | null,
    addMetadata: boolean,
    rememberOriginalUrls: boolean,
    contentSecurityPolicy: ContentSecurityPolicy | null,
    now: Date,
    fetchResource?: Fetchy,
    processSubresource: ProcessSubresourceCallback,
    newUrlForResource: NewUrlForResourceCallback,
    getDocInFrame?: (frameElement: FrameElement) => Document | null, // TODO expose to user & test.
    glob?: typeof window, /* global window */
}
