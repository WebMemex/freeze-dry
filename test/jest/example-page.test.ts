import fs from 'fs'
import jsdom from 'jsdom'
import jestFetchMock from 'jest-fetch-mock' // magically polyfills Response, Request, ...
import { dataURLToBlob } from 'blob-util'

import freezeDry from '../../src/index'
import { NewUrlForResourceCallback, ProcessSubresourceCallback } from '../../src/types'

const fetch = jestFetchMock
Object.assign(global, { fetch })
beforeEach(() => {
    fetch.mockImplementation(mockFetch)
})

jest.useFakeTimers()

const TEST_PAGE_URL = 'https://example.com/pages/example.html'

const noNonsenseConfig = {
    charsetDeclaration: '',
    addMetadata: false,
    keepOriginalAttributes: false,
    setContentSecurityPolicy: false,
}

test('should freeze-dry a trivial example', async () => {
    const inputHtml = `<html><head></head><body><h1>tada!</h1></body></html>`
    const doc = await makeDom(inputHtml, TEST_PAGE_URL)
    const result = await freezeDry(doc, noNonsenseConfig)
    expect(result).toBe(inputHtml)
})

test('should freeze-dry an example with an image', async () => {
    const inputHtml = `<html><head></head><body><img src="/imgs/8x8.png"></body></html>`
    const doc = await makeDom(inputHtml, TEST_PAGE_URL)
    const result = await freezeDry(doc, noNonsenseConfig)
    expect(result).toBe(`<html><head></head><body><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAICAYAAADED76LAAAAEUlEQVQYlWNgaGD4jxePDAUAE0dfwfSqhOEAAAAASUVORK5CYII="></body></html>`)
})

test('should freeze-dry an example page as expected', async () => {
    const doc = await getExampleDoc()

    // Run freeze-dry, while passing a fixed date for reproducability.
    const result = await freezeDry(doc, { now: new Date(1534615340948) })

    expect(result).toMatchSnapshot() // compares to (or creates) snapshot in __snapshots__ folder
})

test('should be idempotent', async () => {
    const doc = await getExampleDoc()
    const dryHtml = await freezeDry(doc, { now: new Date(1534615340948) })
    const docUrl = 'https://url.should.be/irrelevant'
    const dryDoc = await makeDom(dryHtml, docUrl)

    // Freeze-dry the freeze-dried page. Adding metadata would of course break idempotency.
    const extraDryHtml = await freezeDry(dryDoc, { addMetadata: false })

    expect(extraDryHtml).toEqual(dryHtml)
})

test.skip('should return the incomplete result after given timeout', async () => {
    const doc = await getExampleDoc()

    // Make fetch never resolve
    fetch.mockImplementation(url => new Promise(resolve => {}))

    const resultP = freezeDry(doc, {
        now: new Date(1534615340948),
        timeout: 2000,
    })
    jest.runAllTimers() // trigger the timeout directly.
    const result = await resultP

    expect(result).toMatchSnapshot()
})

test('should use the given docUrl', async () => {
    const docUrl = TEST_PAGE_URL
    const docHtml = await (await fetch(docUrl)).text()
    // This time, we use a DOMParser, and create a Document that lacks a URL.
    const doc = new DOMParser().parseFromString(docHtml, 'text/html')

    const result = await freezeDry(doc, { docUrl })

    const dryDoc = await makeDom(result)
    expect(dryDoc.querySelector('a').getAttribute('href'))
        .toEqual('https://example.com/pages/something')
    expect(dryDoc.querySelector('img').getAttribute('src'))
        .toMatch(/^data:/)
})

test('should respect the addMetadata option', async () => {
    const testWithOption = async (addMetadata?) => {
        const doc = await getExampleDoc()

        const result = await freezeDry(doc, { addMetadata })

        const dryDoc = await makeDom(result)
        expect(dryDoc.querySelector('link[rel=original]')).not.toBeNull()
        expect(dryDoc.querySelector('meta[http-equiv=Memento-Datetime]')).not.toBeNull()
    }
    await expect(testWithOption(true)).resolves.toEqual(undefined)
    await expect(testWithOption(false)).rejects.toEqual(expect.anything())
    await expect(testWithOption()).resolves.toEqual(undefined) // option should default to true
})

test('should respect the keepOriginalAttributes option', async () => {
    const testWithOption = async (keepOriginalAttributes?) => {
        const doc = await getExampleDoc()

        const result = await freezeDry(doc, { keepOriginalAttributes })

        const dryDoc = await makeDom(result)
        expect(dryDoc.querySelector('img[data-original-src]')).not.toBeNull()
    }
    await expect(testWithOption(true)).resolves.toEqual(undefined)
    await expect(testWithOption(false)).rejects.toEqual(expect.anything())
    await expect(testWithOption()).resolves.toEqual(undefined) // option should default to true
})

test('should respect the setContentSecurityPolicy option', async () => {
    const testWithOption = async (setContentSecurityPolicy?) => {
        const doc = await getExampleDoc()

        const result = await freezeDry(doc, { setContentSecurityPolicy })

        const dryDoc = await makeDom(result)
        expect(dryDoc.querySelector('meta[http-equiv=Content-Security-Policy]')).not.toBeNull()
    }
    await expect(testWithOption(true)).resolves.toEqual(undefined)
    await expect(testWithOption(false)).rejects.toEqual(expect.anything())
    await expect(testWithOption()).resolves.toEqual(undefined) // option should default to true
})

test('should use the custom fetchResource function', async () => {
    const now = new Date(1545671350764)
    const doc = await getExampleDoc()
    const expectedResult = await freezeDry(doc, { now })

    fetch.mockClear() // Clear the invocation count.
    const fetchResource = jest.fn(mockFetch) // Create a second mock using the same implementation.

    const result = await freezeDry(doc, { now, fetchResource })

    // We should have got the same result as usual, without the global fetch having been invoked.
    expect(result).toEqual(expectedResult)
    expect(fetchResource).toHaveBeenCalled()
    expect(fetch).not.toHaveBeenCalled()
})

test('should work if the custom fetchResource function returns a simple object', async () => {
    const now = new Date(1545671350764)
    const doc = await getExampleDoc()
    const expectedResult = await freezeDry(doc, { now })

    // A fetch-like function, that returns not a Response but a plain object with a blob and a url.
    async function fetchResource(...args) {
        const response = await fetch(...args)
        return {
            url: response.url,
            blob: await response.blob(),
        }
    }

    const result = await freezeDry(doc, { now, fetchResource })

    expect(result).toEqual(expectedResult)
})

test('should use interfaces of a custom global object', async () => {
    const now = new Date(1545671350764)
    const doc = await getExampleDoc()
    const expectedResult = await freezeDry(doc, { now })

    // Create a clone of the window object, replacing some function freezeDry will invoke.
    const glob = Object.create(doc.defaultView)
    glob.btoa = jest.fn(glob.btoa)

    const result = await freezeDry(doc, { now, glob })
    expect(result).toEqual(expectedResult)
    expect(glob.btoa).toHaveBeenCalled()
})

test('should use custom processSubresource', async () => {
    const doc = await getExampleDoc()
    const processSubresource: ProcessSubresourceCallback = (link, recurse) => {
        link.target = 'about:invalid'
    }
    const result = await freezeDry(doc, { processSubresource })
    expect([...result.matchAll(/about:invalid/g)]).toHaveLength(5)
})

test('should use custom newUrlForResource', async () => {
    const doc = await getExampleDoc()
    const newUrlForResource: NewUrlForResourceCallback = resource => 'about:invalid'
    const result = await freezeDry(doc, { newUrlForResource })
    expect([...result.matchAll(/about:invalid/g)]).toHaveLength(5)
})


//////////////// Helper functions ////////////////

async function getExampleDoc(): Promise<Document> {
    const docUrl = TEST_PAGE_URL
    const docHtml = await (await fetch(docUrl)).text()
    const doc = await makeDom(docHtml, docUrl)
    return doc
}

async function makeDom(docHtml: string, docUrl: string = undefined): Promise<Document> {
    const dom = new jsdom.JSDOM(docHtml, {
        url: docUrl,
        resources: new MockResourceLoader(),
    })

    // Make code ‘inside’ jsdom use these ‘outside’ implementations to simplify their interaction.
    // (note we could probably enable mockFetch to use ‘inside’ implementations instead, but at
    // time of writing jsdom lacks an implementation of Response)
    dom.window.fetch = fetch
    dom.window.Blob = Blob
    dom.window.FileReader = FileReader

    const doc = dom.window.document

    // Wait until JSDOM has completed loading the page (including frames).
    await new Promise<void>(resolve => {
        if (doc.readyState === 'complete') {
            resolve()
        } else {
            doc.addEventListener('load', resolve)
        }
    })

    return doc
}

class MockResourceLoader extends jsdom.ResourceLoader {
    async fetch(url: string, options) {
        const response = await mockFetch(url)
        const text = await response.text()
        return Buffer.from(text)
    }
}

// A fetch function that reads the subresources from local files.
async function mockFetch(url: string) {
    const websiteOrigin = 'https://example.com'
    const basedir = __dirname + '/../example-website'

    // We simply use the file extension to determine the mime type.
    const extensionTypes = {
        'html': 'text/html',
        'css': 'text/css',
        'png': 'image/png',
        'woff': 'font/woff',
    }

    if (url.startsWith(websiteOrigin)) {
        const path = url.slice(websiteOrigin.length,)
        const content = fs.readFileSync(basedir + path)
        const mimeType = extensionTypes[url.split('.').reverse()[0]]
        const blob = new Blob([content], { type: mimeType })
        const response = new Response(blob, { status: 200 })
        Object.assign(response, { url })
        return response
    }

    if (url.startsWith('data:')) {
        const blob = await dataURLToBlob(url)
        const response = new Response(blob, { status: 200 })
        Object.assign(response, { url })
        return response
    }

    throw new Error(`Trying to fetch unknown URL: ${url}`)
}
