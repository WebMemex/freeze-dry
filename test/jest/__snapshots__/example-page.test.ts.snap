// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should freeze-dry an example page as expected 1`] = `
"<!DOCTYPE html>
<html><head><meta charset=\\"utf-8\\"><meta http-equiv=\\"Content-Security-Policy\\" content=\\"default-src 'none'; img-src data:; media-src data:; style-src data: 'unsafe-inline'; font-src data:; frame-src data:\\"><meta http-equiv=\\"Memento-Datetime\\" content=\\"Sat, 18 Aug 2018 18:02:20 GMT\\"><link rel=\\"original\\" href=\\"https://example.com/pages/example.html\\">
        
        
        <link rel=\\"stylesheet\\" href=\\"data:text/css;base64,QGltcG9ydCAnZGF0YTp0ZXh0L2NzcztiYXNlNjQsYUhSdGJDQjdDaUFnSUNCaVlXTnJaM0p2ZFc1a09pQmpiM1psY2lCMWNtd29aR0YwWVRwcGJXRm5aUzl3Ym1jN1ltRnpaVFkwTEdsV1FrOVNkekJMUjJkdlFVRkJRVTVUVldoRlZXZEJRVUZCUlVGQlFVRkNRMEZaUVVGQlFXWkdZMU5LUVVGQlFVUlZiRVZSVmxGSmJWZE9aMkZIUkRSRWQwRkRhRUZIUVRKR1NtUnBVVUZCUVVGQ1NsSlZOVVZ5YTBwbloyYzlQU2tLZlFvPSc7CgpAZm9udC1mYWNlIHsKICAgIGZvbnQtZmFtaWx5OiAnTXkgRm9udCc7CiAgICBzcmM6IHVybCgnZGF0YTpmb250L3dvZmY7YmFzZTY0LGQwOUdSZ0FCQUFBQUFBeVFBQk1BQUFBQUZJUUFBUUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFCR1JsUk5BQUFNQ0FBQUFCd0FBQUFjWHo0cmFFZEVSVVlBQUF3a0FBQUFGUUFBQUJZQUVRQUNSMUJQVXdBQUREd0FBQUFnQUFBQUlFUjJUSFZIVTFWQ0FBQU1YQUFBQUNrQUFBQXF1UHE0OUU5VEx6SUFBQU53QUFBQVFnQUFBRloxZkcyelkyMWhjQUFBQTdRQUFBQXhBQUFBUEFCMEFGeGpkblFnQUFBSzVBQUFBQ01BQUFCR0VSTVd5R1p3WjIwQUFBUG9BQUFHY0FBQURXMDVHbzU4WjJGemNBQUFEQUFBQUFBSUFBQUFDQUFBQUJCbmJIbG1BQUFCcUFBQUFUd0FBQUZpZGVib0ltaGxZV1FBQUFNTUFBQUFOZ0FBQURidjdkNEVhR2hsWVFBQUExQUFBQUFnQUFBQUpBMWxCVDlvYlhSNEFBQURSQUFBQUF3QUFBQU1DbGdBcm14dlkyRUFBQU1FQUFBQUNBQUFBQWdBQUFDeGJXRjRjQUFBQXVRQUFBQWdBQUFBSUFFeEFnNXVZVzFsQUFBTENBQUFBT1FBQUFIQUk0czlEM0J2YzNRQUFBdnNBQUFBRXdBQUFDRC9MZ0FVY0hKbGNBQUFDbGdBQUFDTUFBQUFtQzB4aHh4M1pXSm1BQUFNaUFBQUFBWUFBQUFHYUN4WmwzaWNOY3ZiU2dKUkdBWGcvLy8zSG1mTW1TMWpPVE9Pc3oza1lVeHhLczBESm1PUlFXZ1gxVVZJRWRWRk55RWgzUmErVE5CVCtCNjlqWlNWOE1GYUxGaEE4QXdGbmxQR01JUUhhQjAxcmthbmpORXhBdEtJSXlNa2huTkFJSVIzSURMSEFKQ0U4N3VieVhXejRSVVUxYXFWQktvQ0EvVFhWTEcySldnMWxWc2hkdjZ4VHR1eU03aGlKZjZLaWdIemZ3OXh0SC9TYmxqTkREYXRrTGpESzY3dzByWXFWSVZIRld2UGFUMDFkMjkzVXFhbUp6Tm10bVpzVTlsTmQzUHlKTC9wTjA2QzZsakt1aUhpMGNMQVZtSUtjVzM1aGFhYkZjSEFjU3FjaUpBek0yOG1pZ25ySU1oTjNSSkhWSFE5MWZXS28rS0dUaFNOYVJrcHp3cnhyR093U0NtOTN6OTg3WVR6L3VQaTNwOWNoREVxdVhKWThTK3JIOHZGcDJ4N3NpY3AzcHU5VE52MVdhLy9OakIwemRCaXlKQWl5U0o4QXhjWkxVWUFBUUFBQUFNQWpnQUZBQUFBQUFBQ0FCNEFNQUNMQUFBQWZ3Rk9BQUFBQUFBQUFBQUFBQUN4QUFFQUFBQUJBQURJendiVlh3ODg5UUFmQ0FBQUFBQUF0Vng2QXdBQUFBRFZ2UmlxQUFEL2JRV2hCdTBBQUFBSUFBSUFBQUFBQUFBQzdBQkVBd0FBQUFSc0FHcDRuR05nWkdCZ1AvaHZIZ01EMjFzR2h2OGJXQmN5QUVWUUFETUFrTFlGc25pY1kyQmt2c000Z1lHVmdZVjFGcXN4QXdPakdvUm1Yc0N3anJtTEFRaFlHT0NBa1FFSkJIdEV1akU0TUNnd09MS2wvVXRqWUdBL3lKZ0VVd01Bd09VS0NnQUFlSnhqWUdCZ1ltQmdZQVppRVNESkNLWlpHRFNBTkJ1UVpnVEtLakE0L3Y4UDVJUHAvdy8vSHdTckFnSUFqQTBJeWdBQUFIaWNyVlpwZHhNM0ZOVjRTVXpJUmtKQ3k3UlVSamlsc2NaUXloTEFRSmlKNDRLN09BSGFHU2p0VE96UWZZRnVkTjhYL0d2ZUpPMDU5QnMvcmZkSnRra2dvYWM5OVFlL0srbEtiOVhUa05DU3hKV3dGa25adUN0R2x4bzBjT2xxU0VkZE9oakZOMlRuU2tpWlV2SlhRUlJFcTZWVzNHS1JSRVFpVUF0cndoRkI3SHZrYUpMeERZOHlXcllsM1d0U2J1YnEya0ZuWjFCcjFXaWdGaFlwVzRxV3I0VkZWWFE3b2FSbUUxUHprU3RwanRGY0ZNblVzcE0ySGNSVWR5VHBNSzhmWnVhOVppaGhUU2VSTk5RTVk4eElYaHRpZEp6UjhkaU5veWh5WVMwTkJTMFN5eUdKQnBQQkN0d0c3V08wcjVIY0hSY3RadHpOaTVVb2FpY1JPZVVvVWlTYTRXb1VlWlRWRXBwenBRUys1SU5tU0hubDA0RHk0VG1vc1VjNXJlQ0piS2Y1RlYveUN2dm9XcHY1bnpKeHJVWFoyU0lXQTltUkhTaElEK2RMQ010U0dEZmRaRGtLVlZTTUpNMWZDckhtY2pDNitqM0theG9NeW1zaVkyTTdnS0h5RlhLay9JUXlLemZJYWNFS3lzOTZOS2dsbXpvTVgzSmlSZklKTkI5SFRJa1hqS2tGdlRZNExJS2FQMXZzWjJ1SDNweTlJWHVLVTRZSkFmeU9aYTJqRXM2a2liQndPUXNrWFJqWnN4TDVWTW1DVmJGem0rMTBBTHVFZTkrMWpadUd0WEZvYmVkUUZ1WGhxbUkwVy9Sb1JLZVpUSTNheVlKSG94cEVLV2trdU1qYkFaUWYwU2lQbGpFYXhjaWpNUnd6YmtJaUVZRVc5TkpZRU10T0xHa01RZk5vWERjdWgybXV2UkFkb0pGVmRkdWpYYnF4RkRZdTJVbTNpUGxKTXoraFV6RWVYQW5UOGZHQW5NU25zVEpYT2FySlQwZjRieFIvNUV3akU5bFNNMHc1ZVBEVzd5Qy9VRHM2VzFUWTFzT3VYZWN0dUR3OEU4R1RPdXl2WTNaenFyWkpZQ3JFcEVLMEFoSm4xeHpITWJtYTFDSVZtZHJsa01hVkwyczBqT0xicVZCd3ZveWgvcytKQ1VlTUNkL3Z4T25FUUpudWxOMzlDTk51K0RaWjltaEtwdzdMYWNTWjVSNmRabGsrcHRNY3k4ZDFtbWU1VjZjRExGMmREcko4UXFjRmxrL3FkQWZMWjdUcXhaMEdZa1JZeVFvNTEvbUNlRFM3WVhHNnYzalRMcFkzTE03MEYyL1p4WDFhMEVqNVAvajNGUHpiQjdzay9HTlpoSDhzOThNL2xncitzVHdBLzFpVzRCL0xHZmpIOG1uNHgvSWcvR09wdGF5YU12VTAxRTdFTWtCdTQ4Q2tFbGRQYzYxV05IbGw4bkFMRCtFQzFPVTJXVlRKbk9JZStraUd5OTRmN3FVMkhTN1V1TkxvMEd5YWQ2WnFJZm9mZS9uc2h2QnN4em1pNVRGaitYTTR6WEpxRCt2RVpkM1NGcDRYMDM4SS9pMmNWWFBwRVdlS2ZUMktlTUNCcmUzSEpVbm1QRHFtSzN1cUhoMy9KeW9LdWdYNkNhUklUSmRrUmRhNUVTQzBGenFkdXFxamM0UjRZOUJvMFIyT084N1Via1I0RGgxcm1uYUJsa01UTFJsYU9pUjgyaEdVVnpzVkpXVzFnek5QYnFiSmlqMlBjc3J2c1NYRjNFdm1sOEwxak14S2R6MHprOTBiK2R4ZkMyalZ5dXhRaTdqWndZUFhOT1llWngrZ1RCQzNGV1dEcEkzbFRKQzR3REgzdHdmM0pEQU5YVjh0SXNjS0doYjVjU29FUmd2TzIwS0pzcDAwaCthQlpPUlJjUG1IVHNXSjdGWEpHSUgvcHUyZzkzV2hFRTcxWWlFeG01L3B4a0pWRWFiVC9TVXFtUFZGVldlbG5NVnFQNFRzakkwMGljdGhSVmJ4ZHJQMTNVbkpkblZUUVFNbGpDNXMvRXl3U2R5cTJydlpVbHp5WnpaWUV2VFNGZk8zeElNdTkxSjhGdjJqd2xGY3BGMUIySFR4a3NwcVZFa3J6bTdjMjNPYlZwZmQ1cWJWK1MzM1BtckhlVTF6NVVjcDlEV2RMSGRnRzljWW5OcVdpb1JXcUlJZGdYR1o2M1BHUmo2aEllVmIxN2xBRmE1UEJUZlBucitBeG9RM3ByZmxYNVowL2YrcVl2YUorMWhWb1ZWdHFKZGkxTFd6aGdZOFYrNUZaUkdqaytXaTZzYWw2MDAvQkhXRVlNcGVlM3lENElaUFZ1Z29idm56Mjh4ZndISE83a2s2Qm54UjB3bUlCa2V4aG5ETFJUeTR2V2k5b0xtZ3FRSDRvbDVEQ3dONENjQmg4TEplYzh4TUU4RE1MREduQnJETUhBYVhtTVBnTW5NWVhOSHI2SVhuZ1Y0QmNneDZWYTg3ZGk0RXNuTVI4eHhHVjVsbjBEWG1HZlFhOHd5Nnpqb0RnTmRaSjRNM1dDZURtSFV5U0ppekNMRENIQVl0NWpCb000ZkJxckhMQjdwaDdHTDBwckdMMFZ2R0xrWnZHN3NZdldQc1l2U3VzWXZSZThZdVJ1OGp4cWY2Q2Z6QWpPZ3M0SWNXbmdQOGlJTnVSdk1ZM2NSYjIrWGNzcEE1SHh1TzArVjhnczJuKzZkK2FrWm14MmNXOG83UExXVDZiWnpUSlh4aElSTyt0SkFKWDRGYjdaLzN0UmtaK2pjV012MWJDNW4rSFhaMkNkOWJ5SVFmTEdUQ2orQ2U2Wi8za3hrWitzOFdNdjBYQzVuK0szWjJDYjlaeUlUZkxXVENIYjIrSTVmcGZkSDZaU3FzVXZaQTgzYnZpZmIrQmwvTTA2cDRuR1B3M3NGd0lpaGlJeU5qWCtRR3hwMGNEQndNeVFVYkdkaWROak5JTVRGb2dWaGI1Wm40T1pnNElHeEZCbkUyTUp2VGFUY0hBd3NEQXhNREo1REg3YlNid1FFSXdUeG1CcGVOS293ZGdSRWJIRG9pTmpLbnVHeFVBL0YyY1RRd01MSTRkQ1NIUklDVVJBSUIwRGhCTmlZZXJSMk0vMXMzc1BSdVpHSncyY3lhd3NiZzRnSUExRTRsNUhpY1kyREFDMjR3M0dEMVpCQmtkV0wxWUpCazRFSm1NeGhCb1F1cklBQitOd1QvQUhpY2ZZL0xhc0pRR0lTL05MYWxteTdFQjhpeVFpL1daVnlKVU9paW0rcXFDMEUwallLYWNJeEs5bjJDUG16WG5mTWJJVW9waCtUTS9KZVpPY0ExSDRRRWpSc2cxM2ZBQVMyeEE3N2dscThLaDh6NHJuQ2pOblBKbUo4S1g5RU1SZ3pJMUMxeExFaVpVeEJ4eDVTMjdsZmRHV3QxSnZwSHZCZ3IyUERFVUxObHJSSXhFczlKcXJtdDZqT3BsdHlyMHFWajU1RytOQXNwWnFiNFpzcnYya3Exc1JSelo2Ny9lUjRkVC8xaTZmN2xFdGR5N0xUbHhKK1ZxRVB2SktIUDc5UlAyRnNlLzZKUEtmbjV0U2srV01LVmpsZVpXdEtsdWx0bFNzN2VhTjYvYzRSRHVIaWNZMkJtQUlQLzJnd2lERmdBQUIzbUFVSUFBQUVBQWYvL0FBOEFBQUFCQUFBQUFOUWttTG9BQUFBQXRWeDZBd0FBQUFEVnZSaXFlSnhqWUdSZ1lPQmhnQUJHSUdRQ2tRQUJGUUFVQUFBQUFBRUFBQUFLQUJ3QUhnQUJSRVpNVkFBSUFBUUFBQUFBLy84QUFBQUFBQUI0bkdOZ1pHQmc0R0pRWTlCZ1lISng4d2xoNE10SkxNbGprR0JnQVlvei9QL1BBQWNBYlpjRlhRQUFBQUFCV1pkb0t3QUEnKSBmb3JtYXQoJ3dvZmYnKTsKfQoKLndpdGgtY3VzdG9tLWZvbnQgewogICAgZm9udC1mYW1pbHk6ICdNeSBGb250JzsKfQoKLmNsYXNzLWluLXN0eWxlc2hlZXQgewogICAgY29sb3I6IHllbGxvdzsKICAgIGJhY2tncm91bmQ6IHVybCgnZGF0YTppbWFnZS9wbmc7YmFzZTY0LGlWQk9SdzBLR2dvQUFBQU5TVWhFVWdBQUFBRUFBQUFCQ0FZQUFBQWZGY1NKQUFBQURVbEVRVlFJbVdOZ2FHRDREd0FDaEFHQTJGSmRpUUFBQUFCSlJVNUVya0pnZ2c9PScpIHJlcGVhdDsKfQo=\\" data-original-href=\\"/style/style.css\\">
        <style>
          .class-in-style-tag {
              background: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVQImWNgaGD4DwAChAGA2FJdiQAAAABJRU5ErkJggg==') repeat;
          }
      </style>
    </head>
    <body style=\\"background: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVQImWNgaGD4DwAChAGA2FJdiQAAAABJRU5ErkJggg==');\\" data-original-style=\\"background: url('/imgs/background.png');\\">
        <h1>Freeze me.</h1>
        <p>
            <a href=\\"https://example.com/pages/something\\">A relative link.</a>
            <img src=\\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAICAYAAADED76LAAAAEUlEQVQYlWNgaGD4jxePDAUAE0dfwfSqhOEAAAAASUVORK5CYII=\\" data-original-src=\\"/imgs/8x8.png\\">
        </p>
        <iframe id=\\"iframe-src\\" src=\\"data:text/html;base64,PCFET0NUWVBFIGh0bWw+CjxodG1sPjxoZWFkPjwvaGVhZD4KPGJvZHk+CiAgICA8aDE+aWZyYW1lIGNvbnRlbnQ8L2gxPgogICAgPGltZyBzcmM9ImRhdGE6aW1hZ2UvcG5nO2Jhc2U2NCxpVkJPUncwS0dnb0FBQUFOU1VoRVVnQUFBQUVBQUFBQkNBWUFBQUFmRmNTSkFBQUFEVWxFUVZRSW1XTmdhR0Q0RHdBQ2hBR0EyRkpkaVFBQUFBQkpSVTVFcmtKZ2dnPT0iIGRhdGEtb3JpZ2luYWwtc3JjPSJpbWFnZS5wbmciPgogICAgPGlmcmFtZSBzcmM9ImRhdGE6dGV4dC9odG1sO2Jhc2U2NCxQQ0ZFVDBOVVdWQkZJR2gwYld3K0NqeG9kRzFzUGp4b1pXRmtQand2YUdWaFpENEtQR0p2WkhrK0NpQWdJQ0E4YURFK2JtVnpkR1ZrSUdsbWNtRnRaU0JqYjI1MFpXNTBQQzlvTVQ0S0NnbzhMMkp2WkhrK1BDOW9kRzFzUGc9PSIgZGF0YS1vcmlnaW5hbC1zcmM9ImlubmVyaW5uZXJwYWdlLmh0bWwiPjwvaWZyYW1lPgogICAgCgoKPC9ib2R5PjwvaHRtbD4=\\" data-original-src=\\"/iframe/innerpage.html\\"></iframe>
    

</body></html>"
`;

exports[`should return the incomplete result after given timeout 1`] = `
"<!DOCTYPE html>
<html><head><meta charset=\\"utf-8\\"><meta http-equiv=\\"Content-Security-Policy\\" content=\\"default-src 'none'; img-src data:; media-src data:; style-src data: 'unsafe-inline'; font-src data:; frame-src data:\\"><meta http-equiv=\\"Memento-Datetime\\" content=\\"Sat, 18 Aug 2018 18:02:20 GMT\\"><link rel=\\"original\\" href=\\"https://example.com/pages/example.html\\">
        
        
        <link rel=\\"stylesheet\\" href=\\"https://example.com/style/style.css\\">
        <style>
          .class-in-style-tag {
              background: url('https://example.com/imgs/background.png') repeat;
          }
      </style>
    </head>
    <body style=\\"background: url('https://example.com/imgs/background.png');\\">
        <h1>Freeze me.</h1>
        <p>
            <a href=\\"https://example.com/pages/something\\">A relative link.</a>
            <img src=\\"https://example.com/imgs/8x8.png\\">
        </p>
        <iframe id=\\"iframe-src\\" src=\\"data:text/html;base64,PCFET0NUWVBFIGh0bWw+CjxodG1sPjxoZWFkPjwvaGVhZD4KPGJvZHk+CiAgICA8aDE+aWZyYW1lIGNvbnRlbnQ8L2gxPgogICAgPGltZyBzcmM9Imh0dHBzOi8vZXhhbXBsZS5jb20vaWZyYW1lL2ltYWdlLnBuZyI+CiAgICA8aWZyYW1lIHNyYz0iZGF0YTp0ZXh0L2h0bWw7YmFzZTY0LFBDRkVUME5VV1ZCRklHaDBiV3crQ2p4b2RHMXNQanhvWldGa1Bqd3ZhR1ZoWkQ0S1BHSnZaSGsrQ2lBZ0lDQThhREUrYm1WemRHVmtJR2xtY21GdFpTQmpiMjUwWlc1MFBDOW9NVDRLQ2dvOEwySnZaSGsrUEM5b2RHMXNQZz09IiBkYXRhLW9yaWdpbmFsLXNyYz0iaW5uZXJpbm5lcnBhZ2UuaHRtbCI+PC9pZnJhbWU+CiAgICAKCgo8L2JvZHk+PC9odG1sPg==\\" data-original-src=\\"/iframe/innerpage.html\\"></iframe>
    

</body></html>"
`;

exports[`should return the incomplete result if aborted 1`] = `
"<!DOCTYPE html>
<html><head><meta charset=\\"utf-8\\"><meta http-equiv=\\"Content-Security-Policy\\" content=\\"default-src 'none'; img-src data:; media-src data:; style-src data: 'unsafe-inline'; font-src data:; frame-src data:\\"><meta http-equiv=\\"Memento-Datetime\\" content=\\"Sat, 18 Aug 2018 18:02:20 GMT\\"><link rel=\\"original\\" href=\\"https://example.com/pages/example.html\\">
        
        
        <link rel=\\"stylesheet\\" href=\\"https://example.com/style/style.css\\">
        <style>
          .class-in-style-tag {
              background: url('https://example.com/imgs/background.png') repeat;
          }
      </style>
    </head>
    <body style=\\"background: url('https://example.com/imgs/background.png');\\">
        <h1>Freeze me.</h1>
        <p>
            <a href=\\"https://example.com/pages/something\\">A relative link.</a>
            <img src=\\"https://example.com/imgs/8x8.png\\">
        </p>
        <iframe id=\\"iframe-src\\" src=\\"data:text/html;base64,PCFET0NUWVBFIGh0bWw+CjxodG1sPjxoZWFkPjwvaGVhZD4KPGJvZHk+CiAgICA8aDE+aWZyYW1lIGNvbnRlbnQ8L2gxPgogICAgPGltZyBzcmM9Imh0dHBzOi8vZXhhbXBsZS5jb20vaWZyYW1lL2ltYWdlLnBuZyI+CiAgICA8aWZyYW1lIHNyYz0iZGF0YTp0ZXh0L2h0bWw7YmFzZTY0LFBDRkVUME5VV1ZCRklHaDBiV3crQ2p4b2RHMXNQanhvWldGa1Bqd3ZhR1ZoWkQ0S1BHSnZaSGsrQ2lBZ0lDQThhREUrYm1WemRHVmtJR2xtY21GdFpTQmpiMjUwWlc1MFBDOW9NVDRLQ2dvOEwySnZaSGsrUEM5b2RHMXNQZz09IiBkYXRhLW9yaWdpbmFsLXNyYz0iaW5uZXJpbm5lcnBhZ2UuaHRtbCI+PC9pZnJhbWU+CiAgICAKCgo8L2JvZHk+PC9odG1sPg==\\" data-original-src=\\"/iframe/innerpage.html\\"></iframe>
    

</body></html>"
`;
